### ISSUE TITLE
There should be a way to set specific mount options for NFS

### ISSUE THREAD

**Issue by pmorie** [kubernetes/kubernetes#17226]
... in order to do things like turn off attribute caching, which is needed when using an NFS volume as storage for a docker registry.

cc @kubernetes/rh-storage @ncdc

**Comment by rootfs** [comment]
@pmorie +1
we need the ability to set nfs versions, security, caching.

should make it more general for other network filesystems too (gluster, cephfs, and upcoming cifs)

**Comment by thockin** [comment]
Is this something users want to see or something that admins would set on
the PV but users would never experience?  If it is workload-specific, it
can't be just PV, but that means it becomes part of Claim, which nullifies
the decoupling we're trying so hard to keep.

**Comment by markturansky** [comment]
either part of Claim or on the persistent-volume plugin?  The latter is where we can set ReadOnly on the actual mount we're attempting, not on the claim.  Perhaps more runtime params on the pv plugin (e.g, use this secret or config for this mount, but not necessarily the next one with the same PV).

**Comment by ncdc** [comment]
For our particular immediate needs, this is so you can tune the settings for the NFS volume used to back our Docker registry replicas, as the default settings result in close to 100% push failures. I see this more as settings an admin would set on a volume for a specific use case such as this one, and not something you'd want end users to have to worry about with PVCs.

**Comment by smarterclayton** [comment]
Runtime params potentially have security options, so for the most part this has to be set by an admin.  You can't for instance allow someone to specify no_root_squash.

**Comment by remoe** [comment]
Is it not possible to configure NFS with enabled CacheFS by using the 'fsc' option? @smarterclayton this option is not security related, right? What are the alternatives when one will use this? Build of a modified nfs storage plugin?

**Comment by smarterclayton** [comment]
Note that I think we should be able to set options, but any options that have security implications we can't add without a way to control that.  I'm in favor of adding options to these.

**Comment by thockin** [comment]
Circling back to older issues: Would some extra fields in the NFSVolumeSource be sufficient for this?  It gets a little more awkward with dynamic provisioning, but static provisioning would work...

**Comment by sosiouxme** [comment]
So could we at least have the non-security-impacting options available?

**Comment by thockin** [comment]
I think at this point, we need someone to try it and see what works.  Does
someone who is invested in this have some time?

**Comment by rootfs** [comment]
let's track it at #24721

**Comment by Cocotton** [comment]
That is definitely something that would be great. The default "local_lock=off" is preventing me from using Consul in my use case scenario. Hit me up if I can help.

**Comment by sferich888** [comment]
cross linking https://github.com/kubernetes/kubernetes/issues/31613

**Comment by robszumski** [comment]
Adding another datapoint, I'm trying to use Amazon EFS, which documents these flags for mounting:

```
sudo mount -t nfs4 -o nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2 fs-abc123.efs.us-east-1.amazonaws.com:/ efs
```

**Comment by kerin** [comment]
We've just run into an issue with this - our Amazon EFS NFS mount in Kubernetes causes fs lock errors on git operations, which is a pretty major problem for us. Are there any known workarounds for setting NFS mount options, however hacky?

**Comment by eparis** [comment]
Closing this issue as this is resolved in 1.6. See:
https://github.com/kubernetes/kubernetes/pull/41906
https://github.com/kubernetes/kubernetes.github.io/pull/2743
